### ISSUE TITLE
False negative R0204 (redefined-variable-type) with conditionals

### ISSUE THREAD

**Issue by allanlewis** [PyCQA/pylint#799]
I have code like:
```py
from itertools import repeat
def a():
    for x in y:
        yield x
if foo:
    bar = a()
else:
    bar = repeat(None)
```
This causes R0204 to be raised:
```
Redefinition of bar type from generator to itertools.repeat
```
As far as I'm concerned, these are both iterables and I just want a dummy iterable (`repeat(None)`) when `foo` is falsey. I quite like R0204 in general, but I do this type of thing in a few places so I'll probably just disable it globally for now.

This is with the following versions:
```
pylint 1.5.4, 
astroid 1.4.4
Python 2.7.10 (default, Sep  8 2015, 17:20:17) 
[GCC 5.1.1 20150618 (Red Hat 5.1.1-4)]
```

**Comment by The-Compiler** [comment]
I think this is covered by #710 and #742, so I'm closing this.

@PCManticore - what's keeping #742 from getting merged? Looks like the failure was unrelated to @lmedioni's PR, but there are conflicts now.

**Comment by allanlewis** [comment]
Agreed, this should be resolved by #742.

**Comment by The-Compiler** [comment]
Whoops, I forgot to actually *do* the closing part :wink:

**Comment by PCManticore** [comment]
@The-Compiler lack of time for a proper review on my side. Things have been pretty hectic in January, but will get better in February.

I'll merge it in a couple of days, the conflict can be solved locally.
